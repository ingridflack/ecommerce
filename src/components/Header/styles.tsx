import styled, { css } from "styled-components";

export const Container = styled.header`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: ${({ theme }) => theme.spacings.big};
`;

export const Cart = styled.div`
  ${({ theme }) => css`
    display: flex;
    cursor: pointer;
    transition: opacity 200ms ease;

    &:hover {
      opacity: 0.6;
    }

    div {
      display: flex;
      flex-direction: column;
      align-items: flex-end;
      margin-right: ${theme.spacings.small};

      strong {
        font-weight: 800;
      }

      span {
        font-size: ${theme.fontSize.small};
      }
    }
  `}
`;
