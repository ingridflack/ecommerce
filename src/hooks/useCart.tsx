import {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useState,
} from "react";
import { IProductItem } from "../interfaces";
import { toast } from "react-toastify";
import { checkProductStock } from "../services";

interface CartProviderProps {
  children: ReactNode;
}

interface UpdateProductAmount {
  productId: string;
  quantity: number;
}

interface CartContextData {
  cart: IProductItem[];
  addProduct: (product: IProductItem) => Promise<void>;
  removeProduct: (productId: string) => void;
  updateProductAmount: ({ productId, quantity }: UpdateProductAmount) => void;
  clearCart: () => void;
}

const CartContext = createContext<CartContextData>({} as CartContextData);

const STORAGE_KEY = "@Ecommerce:cart";

export function CartProvider({ children }: CartProviderProps): JSX.Element {
  const [cart, setCart] = useState<IProductItem[]>([]);

  useEffect(() => {
    const storagedCart =
      typeof localStorage === "undefined"
        ? null
        : localStorage.getItem(STORAGE_KEY);

    setCart(storagedCart ? JSON.parse(storagedCart) : []);
  }, []);

  const clearCart = () => {
    setCart([]);
  };

  const addProduct = async (product: IProductItem) => {
    try {
      const updatedCart = [...cart];

      const productExists = updatedCart.find(({ id }) => product.id === id);

      const { data: productStock } = await checkProductStock(product.id);

      const stockAmount = productStock.qty_stock;
      const currentAmount = productExists ? productExists.quantity : 0;

      const amount = currentAmount + 1;

      if (amount > stockAmount) {
        toast.error("Quantidade solicitada fora de estoque");
        return;
      }

      if (productExists) {
        productExists.quantity = amount;
      } else {
        const newProduct = {
          ...product,
          quantity: 1,
        };

        updatedCart.push(newProduct);
      }

      setCart(updatedCart);

      window.localStorage.setItem(STORAGE_KEY, JSON.stringify(updatedCart));
    } catch {
      toast.error("Erro ao adicionar o produto");
    }
  };

  const removeProduct = (productId: string) => {
    try {
      const updatedCart = [...cart];

      const productIndex = updatedCart.findIndex(
        (product) => product.id === productId
      );

      if (productIndex >= 0) {
        updatedCart.splice(productIndex, 1);
        setCart(updatedCart);
      } else {
        throw Error();
      }

      window.localStorage.setItem(STORAGE_KEY, JSON.stringify(updatedCart));
    } catch {
      toast.error("Erro na remoção do produto");
    }
  };

  const updateProductAmount = async ({
    productId,
    quantity,
  }: UpdateProductAmount) => {
    try {
      if (quantity <= 0) return;

      const { data } = await checkProductStock(productId);

      const { qty_stock: stockAmount } = data;

      if (quantity > stockAmount) {
        toast.error("Quantidade solicitada fora de estoque");
        return;
      } else {
        const updatedCart = [...cart];

        const productExists = updatedCart.find(
          (product) => product.id === productId
        );

        if (productExists) {
          productExists.quantity = quantity;

          setCart(updatedCart);
          window.localStorage.setItem(STORAGE_KEY, JSON.stringify(updatedCart));
        } else {
          throw Error();
        }
      }
    } catch {
      toast.error("Erro na alteração de quantidade do produto");
    }
  };

  return (
    <CartContext.Provider
      value={{
        cart,
        addProduct,
        removeProduct,
        updateProductAmount,
        clearCart,
      }}
    >
      {children}
    </CartContext.Provider>
  );
}

export function useCart(): CartContextData {
  const context = useContext(CartContext);

  return context;
}
