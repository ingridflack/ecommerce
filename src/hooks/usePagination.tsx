import { useCallback, useMemo, useState } from "react";
import { IProductItem } from "../interfaces";

const INITIAL_PAGINATION = { page: 0, limit: 25 };

const usePagination = (productList: IProductItem[], search: string) => {
  const [pagination, setPagination] = useState({ ...INITIAL_PAGINATION });

  const pageCount = useMemo(
    () => Math.ceil(productList.length / pagination.limit),
    [productList.length, pagination.limit]
  );

  const filteredList = useMemo(
    () =>
      productList.filter(({ name }) =>
        name.toLowerCase().includes(search.toLowerCase())
      ),
    [productList, search]
  );

  const paginatedList = useMemo(() => {
    const { limit, page } = pagination;
    const initialStep = limit * page;

    return filteredList.slice(initialStep, initialStep + limit);
  }, [pagination, filteredList]);

  const onPageChange = useCallback(({ selected }: { selected: number }) => {
    setPagination((prev) => ({ ...prev, page: selected }));
  }, []);

  console.log({ productList, filteredList, paginatedList });

  return { paginatedList, pageCount, onPageChange };
};

export default usePagination;
