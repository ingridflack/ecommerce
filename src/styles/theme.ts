const theme = {
  colors: {
    background: "#fbfefd",

    gray: {
      white: "#FFFFFF",
      platinum: "#e1e1e6",
      silver: "#a8a8b3",
      dark: "#141414",
    },

    primary: "#52b591",
    primaryDark: "#49a282",
  },
  fontSize: {
    small: "0.75rem",
    medium: "0.875rem",
    normal: "1rem",
    large: "1.5rem",
  },
  spacings: {
    tiny: "0.4rem",
    small: "0.8rem",
    medium: "1rem",
    big: "2rem",
  },
} as const;

export type Theme = typeof theme;

export default theme;
