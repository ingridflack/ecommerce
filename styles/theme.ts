const theme = {
  colors: {
    background: "#fbfefd",

    gray: {
      white: "#FFFFFF",
      platinum: "#e1e1e6",
      silver: "#a8a8b3",
      dark: "#141414",
    },

    primary: "#52b591",
  },
  fontSize: {
    small: "0.75rem",
    medium: "0.875rem",
    normal: "1rem",
    large: "1.5rem",
  },
} as const;

export type Theme = typeof theme;

export default theme;
