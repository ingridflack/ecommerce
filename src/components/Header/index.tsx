import Link from "next/link";
import { MdShoppingBasket } from "react-icons/md";

import Image from "next/image";

import * as S from "./styles";
import { useCart } from "../../hooks/useCart";

const Header = () => {
  const { cart } = useCart();
  const cartSize = cart.length;

  const renderCartSize = () => (
    <span data-testid="cart-size">
      {cartSize ?? "-"} {cartSize === 1 ? "item" : "itens"}
    </span>
  );

  return (
    <S.Container>
      <Link href="/" passHref>
        <a>
          <Image
            src="/assets/logo.svg"
            alt="Logotipo da Shopper"
            height={41}
            width={156}
          />
        </a>
      </Link>

      <Link href="/cart">
        <S.Cart>
          <div>
            <strong>Meu carrinho</strong>
            {renderCartSize()}
          </div>
          <MdShoppingBasket size={36} />
        </S.Cart>
      </Link>
    </S.Container>
  );
};

export default Header;
