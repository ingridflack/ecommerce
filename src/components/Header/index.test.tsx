import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import Header from ".";
import { useCart } from "../../hooks/useCart";
import theme from "../../styles/theme";

jest.mock("../../hooks/useCart");

describe("Header", () => {
  it("should match the snapshot", () => {
    useCart.mockImplementation(() => ({
      cart: [
        {
          quantity: 2,
          id: 1,
          price: 179.9,
          title: "Produto teste",
        },
      ],
    }));

    const { container } = render(
      <ThemeProvider theme={theme}>
        <Header />
      </ThemeProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it("should be able to render the amount of products added to cart - singular", () => {
    useCart.mockImplementation(() => ({
      cart: [
        {
          quantity: 2,
          id: 1,
          price: 179.9,
          title: "Produto teste",
        },
      ],
    }));

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Header />
      </ThemeProvider>
    );

    const cartSizeCounter = getByTestId("cart-size");
    expect(cartSizeCounter.innerHTML).toBe("1 item");
  });

  it("should be able to render the amount of products added to cart - plural", () => {
    useCart.mockImplementation(() => ({
      cart: [
        {
          quantity: 2,
          id: 1,
          price: 179.9,
          title: "Produto teste",
        },
        {
          quantity: 3,
          id: 2,
          price: 19.9,
          title: "Produto teste 2",
        },
      ],
    }));

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Header />
      </ThemeProvider>
    );

    const cartSizeCounter = getByTestId("cart-size");
    expect(cartSizeCounter.innerHTML).toBe("2 itens");
  });
});
