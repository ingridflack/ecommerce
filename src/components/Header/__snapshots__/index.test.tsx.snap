// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match the snapshot 1`] = `
<div>
  <header
    class="sc-d0070584-0 hpDyUT"
  >
    <a
      href="/"
    >
      <span
        style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
      >
        <span
          style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27156%27%20height=%2741%27/%3e"
            style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
          />
        </span>
        <img
          alt="Logotipo da Shopper"
          data-nimg="intrinsic"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
    </a>
    <div
      class="sc-d0070584-1 bBZIim"
    >
      <div>
        <strong>
          Meu carrinho
        </strong>
        <span
          data-testid="cart-size"
        >
          1
           
          item
        </span>
      </div>
      <svg
        fill="currentColor"
        height="36"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="36"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M17.21 9l-4.38-6.56a.993.993 0 00-.83-.42c-.32 0-.64.14-.83.43L6.79 9H2c-.55 0-1 .45-1 1 0 .09.01.18.04.27l2.54 9.27c.23.84 1 1.46 1.92 1.46h13c.92 0 1.69-.62 1.93-1.46l2.54-9.27L23 10c0-.55-.45-1-1-1h-4.79zM9 9l3-4.4L15 9H9zm3 8c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2z"
        />
      </svg>
    </div>
  </header>
</div>
`;
